### PR TITLE
f64/f32: dedicated fused ConvolveValidMaxAbs kernels (peak-detection speedup)

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Follow-up to #139. `ConvolveValidMaxAbs` / `ConvolveValidMaxAbsMulti` were shipped in v1.4.0-rc.1 as a Go-level fusion that looped over the dispatched `dotProduct`. A downstream true-peak consumer measured only ~3-5% over their old `ConvolveValid`-into-a-buffer + scalar abs-max scan, because that fusion removes the output store and the second scan but keeps the per-output dispatch/slice-header overhead, which dominates the MAC work for short kernels.

This replaces the Go-level fusion with a dedicated fused assembly kernel per backend, modeled on the existing `ConvolveDecimate` kernel (with a stride of one): the kernel coefficients stay resident across output positions and the abs-max is folded into each window's reduce, so the per-output call/dispatch/slice-header overhead is gone too.

- f64: `convolveValidMaxAbsAVX` (FMA), `convolveValidMaxAbsSSE2`, `convolveValidMaxAbsNEON`
- f32: `convolveValidMaxAbsAVX` (FMA), `convolveValidMaxAbsSSE`, `convolveValidMaxAbsNEON`
- AVX-512 reuses the AVX2 kernel; FMA-less AVX CPUs use the SSE kernel; ARM64 uses NEON for kLen >= 2 (f64) / >= 4 (f32), Go below that. The Go-level fusion remains the fallback.

**Bit-exactness:** each kernel's inner dot accumulation is byte-identical to the matching `convolveDecimate*` / `dotProduct*` kernel, so the peak stays bit-identical to `max|ConvolveValid output|`. The existing exact-equality tests (and a new direct per-kernel test asserting each kernel equals `max|its per-window dotProduct|`) confirm this; no tolerance was introduced.

**Speedup** (fused vs the rc.1 Go-level fusion / materialized baseline), 32-tap kernels:

| shape | f64 | f32 |
| --- | --- | --- |
| sig 4096, AVX2 | ~1.8x | ~1.6x |
| sig 48000, AVX2 | ~2.2x | ~1.4x |
| sig 4096/48000, NEON (Pi 5) | - | ~1.5x |

Largest win at short kernels, where per-output overhead dominates.

## Related Issues

Relates to #138

## Test Plan

- [x] Existing exact-equality parity tests now exercise the fused kernels via dispatch
- [x] New direct per-kernel test: each fused kernel == `max|matching per-window dotProduct|` (covers SSE/SSE2 on an AVX2 host)
- [x] Edge cases, mismatched-kernel panic, `AllocsPerRun == 0`, fuzz (exact)
- [x] FMA-routing test: the FMA-based AVX kernel falls back to SSE on FMA-less CPUs
- [x] Fixed a pre-existing dispatch-snapshot gap that leaked the fused impl across init-routing tests
- [x] asmcheck validates the NEON `WORD`s (all reused from the validated `convolveDecimate` kernels)
- [x] Full suite green on amd64 (SSE2/AVX2/Go) and a Raspberry Pi 5 (NEON); `golangci-lint run ./...` clean

## Summary by CodeRabbit

* **New Features**
  * Expanded the fused “valid convolution + max-absolute peak” operation with optimized SIMD implementations across amd64 (SSE/AVX variants) and arm64 (NEON).

* **Tests**
  * Added new scalar-oracle validation coverage plus amd64 correctness checks, including zero-allocation assertions where supported.

* **Benchmarks**
  * Introduced benchmarks comparing fused performance against a Go fallback across multiple signal/kernel sizes.

* **Documentation**
  * Clarified fused peak behavior (including bit-identical expectations) and updated performance wording for short-kernel ranges.

### CODE REVIEW

File: f32/f32_arm64.s [hunk 1]
```diff
@@ -2920,6 +2920,89 @@ cd_neon_store:
...
+    CBZ R4, cvma_neon_reduce
+
+    // Reduce vector FIRST before scalar ops (scalar ops zero upper V bits).
+    WORD $0x6E20D400             // FADDP V0.4S, V0.4S, V0.4S
+    WORD $0x7E30D800             // FADDP S0, V0.2S
+
+cvma_neon_scalar:
+    FMOVS (R0), F2
+    FMOVS (R1), F4
+    FMADDS F4, F0, F2, F0         // F0 = F2 * F4 + F0
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_arm64.s:2984]: This is a false positive. `FMADDD F4, F0, F2, F0` does compute `F0 = F2*F4 + F0` in Go's ARM64 assembler, not `F0 = F2 + F0*F4`. Go's Plan9 operand convention here is not the ARM architectural `FMADD Dd,Dn,Dm,Da` order you assumed.

I verified empirically on real aarch64 hardware (Raspberry Pi 5): a probe running exactly `FMADDD F4, F0, F2, F0` with F0=10, F2=3, F4=2 returns 16 (= 3*2+10), not 23 (= 3+10*2). This instruction is also copied verbatim from the existing, shipped `convolveDecimateNEON` scalar tail. And the kernel now passes an independent pure-Go `math.FMA` oracle test on the Pi 5 for kLen values that exercise this exact scalar-tail path (kl=5,13,17,31,33); a wrong operand order would make the result grossly off, so the test would fail. No change.

File: f64/f64_arm64.s [hunk 2]
```diff
@@ -1728,6 +1728,80 @@ cd_neon_store:
...
+
+cvma_neon_rem1:
+    AND $1, R3, R4
+    CBZ R4, cvma_neon_reduce
+
+    // Reduce vector FIRST before scalar op (scalar ops zero upper V bits).
+    WORD $0x7E70D800           // FADDP D0, V0.2D
+    FMOVD (R0), F2
+    FMOVD (R1), F4
+    FMADDD F4, F0, F2, F0        // F0 = F2 * F4 + F0
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_arm64.s:1789]: This is a false positive. `FMADDD F4, F0, F2, F0` does compute `F0 = F2*F4 + F0` in Go's ARM64 assembler, not `F0 = F2 + F0*F4`. Go's Plan9 operand convention here is not the ARM architectural `FMADD Dd,Dn,Dm,Da` order you assumed.

I verified empirically on real aarch64 hardware (Raspberry Pi 5): a probe running exactly `FMADDD F4, F0, F2, F0` with F0=10, F2=3, F4=2 returns 16 (= 3*2+10), not 23 (= 3+10*2). This instruction is also copied verbatim from the existing, shipped `convolveDecimateNEON` scalar tail. And the kernel now passes an independent pure-Go `math.FMA` oracle test on the Pi 5 for kLen values that exercise this exact scalar-tail path (kl=5,13,17,31,33); a wrong operand order would make the result grossly off, so the test would fail. No change.

File: f32/convolve_maxabs_amd64_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		for _, c := range cfgs {
+			s, k := mkPair(c.sl, c.kl)
+			got := convolveValidMaxAbsAVX(s, k)
+			want := refConvMaxAbsDot(s, k, dotProductAVX)
+			if got != want {
+				t.Errorf("sl=%d kl=%d: AVX=%v want %v", c.sl, c.kl, got, want)
+			}
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/convolve_maxabs_amd64_test.go:85]: Added in cf7fcf9. Both direct-kernel test files now assert `AllocsPerRun == 0` for the SSE/SSE2 and AVX kernels (the public `ConvolveValidMaxAbs`/`Multi` no-alloc assertions already existed in convolve_maxabs_test.go).
